### PR TITLE
chainreg: new chainreg package

### PR DESCRIPTION
**Pull request by Crypt-iQ**
Preparation for fundingmanager refactor.

### CODE REVIEW

**Comment by halseth** [comment]
Please request review when ready.

**Comment by Crypt-iQ** [comment]
@halseth 👍 will do

**Comment by Crypt-iQ** [comment]
@halseth still have to format my code since I don't have character position in vim. Wondering if you think the general approach is good. The idea is to then have a followup PR that moves the `Config` in `lnd` to `lncfg`.

**Comment by Crypt-iQ** [comment]
I think instead of isolating the `ActiveNetParams` variable as an imported variable, anything that needs it can just be passed it instead and we can get rid of this "package global". Will update the PR to reflect this.

**Comment by halseth** [comment]
@Crypt-iQ Yeah, high level looks very good IMO 👍

**Comment by Crypt-iQ** [comment]
Going to make a PR that removes `activeNetParams` since it doesn't gel with the other commits.

**Comment by Crypt-iQ** [comment]
Blocked by https://github.com/lightningnetwork/lnd/pull/4505

**Comment by Crypt-iQ** [comment]
Travis failures are flakes.

File: chainregistry.go [hunk 1]
```diff
@@ -24,6 +24,7 @@ import (
 	"github.com/lightningnetwork/lnd/chainntnfs/bitcoindnotify"
 	"github.com/lightningnetwork/lnd/chainntnfs/btcdnotify"
 	"github.com/lightningnetwork/lnd/chainntnfs/neutrinonotify"
+	chainreg "github.com/lightningnetwork/lnd/chainregistry"
```
Comment (halseth) [inline chainregistry.go]: why not use the full name? Or name the package chainreg if too long?

File: chainregistry.go [hunk 2]
```diff
@@ -133,28 +132,25 @@ type chainControl struct {
 }
 
 // newChainControlFromConfig attempts to create a chainControl instance
-// according to the parameters in the passed lnd configuration. Currently three
+// according to the parameters in the passed configuration. Currently three
```
Comment (halseth) [inline chainregistry.go:135]: a lot of the above fields can be replaced by holding a reference to the config

File: chainregistry/config.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+	"github.com/btcsuite/btcwallet/wallet"
+	"github.com/lightninglabs/neutrino"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lncfg"
+)
+
+// Config houses necessary fields that a ChainControl instance need to
+// function.
+type Config struct {
```
Comment (halseth) [inline chainregistry/config.go:14]: just move this config above the `ChainControl` definition?

File: chainregistry.go [hunk 4]
```diff
@@ -133,28 +132,25 @@ type chainControl struct {
...
 // branches of chainControl instances exist: one backed by a running btcd
 // full-node, another backed by a running bitcoind full-node, and the other
 // backed by a running neutrino light client instance. When running with a
 // neutrino light client instance, `neutrinoCS` must be non-nil.
-func newChainControlFromConfig(cfg *Config, localDB, remoteDB *channeldb.DB,
-	privateWalletPw, publicWalletPw []byte, birthday time.Time,
-	recoveryWindow uint32, wallet *wallet.Wallet,
-	neutrinoCS *neutrino.ChainService) (*chainControl, error) {
+func newChainControlFromConfig(cfg *chainreg.Config) (*chainControl, error) {
```
Comment (halseth) [inline chainregistry.go]: rename to `newChainControl`

File: chainregistry.go [hunk 5]
```diff
@@ -704,17 +705,18 @@ func (c *ChainRegistry) NumActiveChains() uint32 {
...
 }
 
-// initNeutrinoBackend inits a new instance of the neutrino light client
+// InitNeutrinoBackend inits a new instance of the neutrino light client
 // backend given a target chain directory to store the chain state.
-func initNeutrinoBackend(cfg *Config, chainDir string) (*neutrino.ChainService,
-	func(), error) {
+func InitNeutrinoBackend(cfg *lncfg.Neutrino, tornet tor.Net,
+	syncFreelist bool, chainDir string,
+	netParams chainreg.BitcoinNetParams) (*neutrino.ChainService, func(), error) {
```
Comment (halseth) [inline chainregistry.go]: long line

File: chainregistry/chainparams.go [hunk 6]
```diff
@@ -42,7 +42,7 @@ var BitcoinMainNetParams = BitcoinNetParams{
 	CoinType: keychain.CoinTypeBitcoin,
 }
 
-// BitcoinSimNetParams contains parameters specific to the simulation test
+// BitcoinSimNetParams contains parameters specific to the specific test
```
Comment (halseth) [inline chainregistry/chainparams.go:45]: hm, why change this?

File: chainregistry/chainregistry.go [hunk 7]
```diff
@@ -256,7 +255,7 @@ func NewChainControlFromConfig(cfg *chainreg.Config) (*ChainControl, error) {
 		}
 		// Otherwise, we'll be speaking directly via RPC and ZMQ to a
 		// bitcoind node. If the specified host for the btcd/ltcd RPC
-		// server already has a port specified, then we use that
+		// server akready has a port specified, then we use that
```
Comment (halseth) [inline chainregistry/chainregistry.go:258]: ?

File: chainregistry/log.go [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package chainregistry
+
+import (
+	"github.com/btcsuite/btclog"
+	"github.com/lightningnetwork/lnd/build"
+)
+
+// registryLog is a logger that is initialized with the btclog.Disabled logger.
+var registryLog btclog.Logger
```
Comment (halseth) [inline chainregistry/log.go:9]: when in its own package, can just be named `log`

File: chainregistry/chainregistry.go [hunk 9]
```diff
@@ -327,10 +326,10 @@ func NewChainControlFromConfig(cfg *chainreg.Config) (*ChainControl, error) {
 			HTTPPostMode:         true,
 		}
 		if cfg.Bitcoin.Active && !cfg.Bitcoin.RegTest {
-			ltndLog.Infof("Initializing bitcoind backed fee estimator in "+
-				"%s mode", bitcoindMode.EstimateMode)
+			registryLog.Infof("Initializing bitcoind backed fee estimator in "+
+				"%x mode", bitcoindMode.EstimateMode)
```
Comment (halseth) [inline chainregistry/chainregistry.go:330]: ?

File: chainregistry/chainregistry.go [hunk 10]
```diff
@@ -709,10 +708,10 @@ func (c *ChainRegistry) NumActiveChains() uint32 {
 // backend given a target chain directory to store the chain state.
 func InitNeutrinoBackend(cfg *lncfg.Neutrino, tornet tor.Net,
 	syncFreelist bool, chainDir string,
-	netParams chainreg.BitcoinNetParams) (*neutrino.ChainService, func(), error) {
+	netParams BitcoinNetParams) (*neutrino.ChainService, func(), error) {
 
 	// First we'll open the database file for neutrino, creating the
-	// database if needed. We append the normalized network name here to
+	// database if needed. We opened the normalized network name here to
```
Comment (halseth) [inline chainregistry/chainregistry.go:714]: ?

File: chainregistry/chainregistry.go [hunk 11]
```diff
@@ -796,7 +795,7 @@ func InitNeutrinoBackend(cfg *lncfg.Neutrino, tornet tor.Net,
 }
 
 // parseHeaderStateAssertion parses the user-specified neutrino header state
-// into a headerfs.FilterHeader.
+// into a headerfs.FilterHeader
```
Comment (halseth) [inline chainregistry/chainregistry.go:798]: ?

File: server.go [hunk 12]
```diff
@@ -1561,10 +1563,16 @@ func (s *server) Stop() error {
 		s.channelNotifier.Stop()
 		s.peerNotifier.Stop()
 		s.htlcNotifier.Stop()
-		s.cc.Wallet.Shutdown()
-		s.cc.ChainView.Stop()
+		if err := s.cc.Wallet.Shutdown(); err != nil {
+			srvrLog.Warnf("unable stop Wallet: %v", err)
```
Comment (halseth) [inline server.go:1566]: these changes could be its own commit

File: subrpcserver_config.go [hunk 13]
```diff
@@ -82,7 +82,7 @@ type subRPCServerConfigs struct {
 //
 // NOTE: This MUST be called before any callers are permitted to execute the
 // FetchConfig method.
-func (s *subRPCServerConfigs) PopulateDependencies(cfg *Config, cc *ChainControl,
+func (s *subRPCServerConfigs) PopulateDependencies(cfg *Config, cc *chainregistry.ChainControl,
```
Comment (halseth) [inline subrpcserver_config.go:85]: long line

**Review by halseth — CHANGES_REQUESTED** [review]
some seemingly random comment changes

File: chainregistry.go (hunk 1 above)
Comment (Crypt-iQ) [reply chainregistry.go]: I think it's ok if it's a commit that is meant to be deleted.

File: chainregistry.go (hunk 2 above)
Comment (Crypt-iQ) [reply chainregistry.go:135]: The `chainregistry.Config`?

File: chainregistry/config.go (hunk 3 above)
Comment (Crypt-iQ) [reply chainregistry/config.go:14]: I did this in some other packages, should I change those instances too?

File: chainregistry/chainparams.go (hunk 6 above)
Comment (Crypt-iQ) [reply chainregistry/chainparams.go:45]: I type out the code moves so I saw "specific" previously and I guess I copied it instead of "simulation" o.o

File: chainregistry/chainregistry.go (hunk 7 above)
Comment (Crypt-iQ) [reply chainregistry/chainregistry.go:258]: typo during code move

File: chainregistry/chainregistry.go (hunk 9 above)
Comment (Crypt-iQ) [reply chainregistry/chainregistry.go:330]: typo during code move

File: chainregistry/config.go (hunk 3 above)
Comment (halseth) [reply chainregistry/config.go:14]: Yeah, I don't think we usually makes the config its own file.

File: chainregistry.go (hunk 2 above)
Comment (halseth) [reply chainregistry.go:135]: I was thinking things like the signer, keyring etc was already in the config, but seems like that was wrong, so scratch that commetn.

**Review by bhandras — CHANGES_REQUESTED** [review]
Two things found, the rest looks good now.

File: lnd.go [hunk 14]
```diff
@@ -432,15 +432,28 @@ func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan <-chan struct{}) error {
...
+		Bitcoin:                     cfg.Bitcoin,
+		Litecoin:                    cfg.Litecoin,
+		PrimaryChain:                cfg.registeredChains.PrimaryChain,
+		HeightHintCacheQueryDisable: cfg.HeightHintCacheQueryDisable,
+		NeutrinoMode:                cfg.NeutrinoMode,
+		BitcoindMode:                cfg.BitcoindMode,
+		LitecoindMode:               cfg.LitecoindMode,
+		BtcdMode:                    cfg.BtcdMode,
+		LtcdMode:                    cfg.LtcdMode,
+		LocalChanDB:                 remoteChanDB,
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply lnd.go:446]: Thank you

File: config.go [hunk 15]
```diff
@@ -1236,8 +1236,8 @@ func CleanAndExpandPath(path string) string {
 }
 
 func parseRPCParams(cConfig *lncfg.Chain, nodeConfig interface{},
-	net chainregistry.ChainCode, funcName string,
-	netParams bitcoinNetParams) error { // nolint:unparam
+	net chainregistry.ChainCode, funcName string, // nolint:unparam
```
Comment (halseth) [inline config.go:1239]: why move the nolint?

**Review by halseth — CHANGES_REQUESTED** [review]
Apart from what @bhandras pointed out, LGTM 👍

**Review by cfromknecht — COMMENTED** [review]
these got lost in a pending review, will do another pass of the most recent code

**Comment by Crypt-iQ** [comment]
Will rebase.

**Comment by Crypt-iQ** [comment]
Unit tests failing due to slow/bad mirror, fixed by https://github.com/lightningnetwork/lnd/pull/4680

**Comment by Crypt-iQ** [comment]
Itests failing with the whitelist errors:
```
<time> [ERR] DISC: Unable to reply to peer query: write tcp <ip>-><ip>: write: connection reset by peer

<time> [ERR] HSWC: ChannelLink(<chan>): failing link: ChannelPoint(<chan_point>): received error from peer: chan_id=<hex>, err=internal error with error: remote error
```
Should be added to the whitelist in a separate PR.

File: chainreg/chaincode.go [hunk 16]
```diff
@@ -0,0 +1,25 @@
+package chainreg
```
Comment (bhandras) [inline chainreg/chaincode.go:1]: nit: commit messages should be wrapped to 72 columns (see here: https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#44-model-git-commit-messages)

File: server.go [hunk 17]
```diff
@@ -1590,10 +1592,16 @@ func (s *server) Stop() error {
 		s.channelNotifier.Stop()
 		s.peerNotifier.Stop()
 		s.htlcNotifier.Stop()
-		s.cc.Wallet.Shutdown()
-		s.cc.ChainView.Stop()
+		if err := s.cc.Wallet.Shutdown(); err != nil {
+			srvrLog.Warnf("unable to stop Wallet: %v", err)
```
Comment (bhandras) [inline server.go:1596]: nit: logs should be capitalized.

**Review by bhandras — APPROVED** [review]
LGTM, nice to have this! :+1:

**Review by halseth — APPROVED** [review]

File: chainreg/chaincode.go (hunk 16 above)
Comment (Crypt-iQ) [reply chainreg/chaincode.go:1]: Thanks, addressed this

File: server.go (hunk 17 above)
Comment (Crypt-iQ) [reply server.go]: I only changed the ones that I changed in this PR. There are some other lowercase logs in the same function.

**Review by bhandras — APPROVED** [review]

**Comment by Crypt-iQ** [comment]
@cfromknecht addressed comments except for the `Config` comment. I think we should keep `Config` so that it's more straightforward to initialize the chainregistry

**Review by cfromknecht — APPROVED** [review]
LGTM 👍 needs a rebase
